### ISSUE TITLE
Running on the Raspberry Pi?

### ISSUE THREAD

**Issue by chrisspen** [mozilla/DeepSpeech#1300]
I suspect this is more of a feature request, but is there any way to run DeepSpeech on the Raspberry Pi?

I've tested DS on a conventional laptop, which has plenty of resources, and although it seems to work well, it consumed about 1.5GB of memory. I've been looking for a speech recognition tool, that aren't reliant on Cloud services, for use on embedded systems like the Raspberry PI. However, as the Pi only as 1GB of total memory, it wouldn't be able to run DS in its default configuration.

Is there any way to shrink DS's memory usage so it can be run on smaller platforms?

**Comment by lissyx** [comment]
This is already something we are working for. Please take discussion on Discourse: https://discourse.mozilla.org/c/deep-speech

Long story short, it's still a long road to be able to run with decent enough performances on RPi3.

**Comment by lissyx** [comment]
You can find binaries out of TaskCluster: https://tools.taskcluster.net/index/project.deepspeech.deepspeech.native_client.master/arm
Python and NodeJS packages are also pushed to PyPI/NPM registries.

**Comment by lissyx** [comment]
Moving to a streaming model (#1275) will help (a lot). Model compiled with `tfcompile` also helps, a lot, but it's more or less experimental (several issues on tensorflow side, and we made no serious benchmark regarding impact of WER). Most of the memory consumption comes from protobuf itself. `mmap`-enabled protobuf helps reducing, but somehow it explodes on RPi3 (crashed with `std::bad_alloc`) and we have not had enough time to debug that so far.

**Comment by lastmjs** [comment]
What about using the GPU on the Raspberry Pi? Also, can you point me to any information about how to use `tfcompile`? I'm trying to squeeze out as much performance as possible on my Raspbeery PI 3

**Comment by lissyx** [comment]
@lastmjs As much as I could gather, it's out of hope to expect reasonable working OpenCL on the RPi3. There's a project on Github, but as much as I remember, it was still in its infancy, and it was already advertised it could not ever be compatible with whole specs, so I don't have big hopes. Also, TensorFlow support for OpenCL is still rough, and I have to blend lot of pieces from everywhere to get something working with recent TensorFlow and ComputeCpp lib (requirement from TensorFlow). It's working, but not decently enough (though we have the `ccpp` branch live).

Regarding `tfcompile`, we have builds online: https://tools.taskcluster.net/index/project.deepspeech.deepspeech.native_client.master/arm_aot but be aware it's very very experimental, more online to make sure we don't break stuff than to be usable so far. Regarding **how** to use it, I would direct you to https://github.com/mozilla/DeepSpeech/blob/31c01db541e6c9d9e8cb86fa766e643ece95c29b/native_client/BUILD#L18-L46 and https://github.com/mozilla/DeepSpeech/blob/31c01db541e6c9d9e8cb86fa766e643ece95c29b/native_client/BUILD#L111-L114

**Comment by edubois** [comment]
I think the Rock64Pro-AI would be a good candidate to try deepspeech on.

**Comment by sbrl** [comment]
OpenCL support would be brilliant for those of us with AMD cards though, as CUDA is an nvidia-specific technology.

**Comment by lissyx** [comment]
@sbrl We know. But it also depends on TensorFlow's support. And so far, it's still not that much upstream.

**Comment by sbrl** [comment]
Ah, I see. I'll track down the issue upstream and subscribe to that then :-)

**Comment by lissyx** [comment]
@sbrl It's possible, we have been focusing on other things lately, especially I worked with ComputeCpp, upstream for TensorFlow with OpenCL and the VC4 driver. Back then, there was an old ComputeCpp-based OpenCL support in TensorFlow, maybe they updated that with what I re-tested recently? It's true the branch I was working on with them for RPi3 GPU and Intel GPU were 1.8~ish.

I already commented that somewhere else, but we have some WIP branch with newer (ComputeCpp fork) TensorFlow and OpenCL-based builds. It was working decently, just stayed as WIP because of the lack of tensorflow's upstream integration. If they did it in 1.8, then we can look at that once we release 0.2.0 with streaming, so likely in a few weeks.
